File: BOJ/♠ 1000~1999/1654_랜선_자르기.py
import sys
input = sys.stdin.readline
num, need = list(map(int, input().split()))
line = []
for i in range(num):
    line.append(int(input()))
start, end = 1, max(line)

while start <= end:
    mid = (start + end) // 2
    sum_ = 0
    for i in range(num):
        sum_ += line[i] // mid
    if sum_ >= need:
        start = mid + 1
    else:
        end = mid - 1

print(end)

File: BOJ/♠ 1000~1999/1654_랜선_자르기.js
const isPossible = (sliceSize) => {
  let lineCount = 0;

  input.forEach((line) => {
    lineCount += Math.floor(line / sliceSize);
  });

  return lineCount >= M;
}

const [NM, ...input] = require('fs')
  .readFileSync(0, 'utf-8')
  .trim()
  .split('\n');

const M = NM.split(' ')[1];

let start = 1;
let end = 2 ** 31 - 1;
let mid;

while (start <= end) {
  mid = Math.floor((start + end) / 2);

  if (isPossible(mid)) {
    start = mid + 1;
  } else {
    end = mid - 1;
  }
}

console.log(end);
